Instant Search: Open overlay while typing if results have loaded

jsnmoon opened the pull request:
Fixes #14612.

#### Changes proposed in this Pull Request:
* Opens search overlay while the user is typing into a site search input if the client has loaded search results from the API.
* Fixes an overlay input focusing bug introduced by one of my hotfixes during one of the input polish PRs.

#### Is this a new feature or does it add/remove features to an existing part of Jetpack?
* No.

#### Testing instructions:
1. Follow [these setup instructions](https://github.com/Automattic/jetpack/blob/instant-search-master/modules/search/instant-search/README.md#testing-instructions) to set up Jetpack Instant Search on your site. Please note that the Jetpack Search widget will need to be added to twice: once to your site sidebar/footer and once to the Jetpack Search Sidebar within the search overlay.

2. Start typing into one of your site search inputs. 

3. Ensure that the overlay appears with your results as you are typing. 

4. Ensure that your browser focus has shifted smoothly from your original site search input to the overlay search input.

5. Close the search overlay and clear one of the site search inputs (e.g. `cmd + a` and `delete`). Ensure that the overlay does not spawn in this instance. 

#### Proposed changelog entry for your changes:
* None.

Commit:
Return promise for SearchApp.getResults

Commit:
Refactor and fix overlay input focus handling

Commit:
Open overlay while typing if results are available

gibrown commented:
Ya, dang it feels too damn fast.

Can we try it with more debouncing on the input typing? It seems to immediately open the overlay as soon as I start typing which feels very different from when you are in the overlay. I wonder if we should wait a bit more for the end user to pause in their typing.

On mobile it is not as jarring and is maybe a better experience.

Bugs I saw:
- when typing on mobile (Chrome on iOS) I lost a character as the overlay was opening. So i typed "rel" and the overlay started opening after "r" and the input box in the overlay ended up with "rl" in it.
- When I hit escape from the overlay it closes the overlay, but does not clear the search box on the main page. If I then delete a character it opens the overlay again. Maybe we should ignore deletions in the input box?

Commit:
Debounce input handler by 350ms

Commit:
Ignore deletion and formatting input events

jsnmoon commented:
> Can we try it with more debouncing on the input typing?

> when typing on mobile (Chrome on iOS) I lost a character as the overlay was opening. 

> Maybe we should ignore deletions in the input box? @gibrown 

I added a 350ms debounce on the input handler, which should also address characters being lost as the overlay is being opened. I also added an exception to ignore all deletion events.

Commit:
Restore support for handling form submissions

gibrown committed:
Lower debounce to 200ms

gibrown commented:
That feels a lot better. I lowered it a bit down to 200ms which felt a little snappier but still not crazy.

Ran into a different bug on mobile though. When typing I no longer lose letters, but my first letter typed into the overlay is getting auto-capitalized because the iOS capitalization is turning on when the new input area gets focus (I guess).

gibrown commented:
It is interesting to convert these delays into words per minute (based on English ave of 4.7 characters per word):
- 350ms => 37 wpm
- 200ms => 64 wpm
- 100ms => 128 wpm

Interesting graphic: https://www.i-programmer.info/images/stories/News/2018/april/A/speed.jpg

But the slower a typist you are, the more utility there is in showing you results sooner since it saves you more time to type the same word (in theory).

jsnmoon commented:
> my first letter typed into the overlay is getting auto-capitalized @gibrown 

Although I'm not able to reproduce this on mobile Safari nor iOS Chrome, I think this behavior is okay to ship.

keoshi reviewed (commented):
Really digging the perceived speed of typing and getting instant search results, this is looking really good. I especially like how seamless it goes from one input to another. However, I thought the overlay would only show in case search results were returned?

If I type `a` on my site no content is found but the overlay is shown despite that:

![image](https://user-images.githubusercontent.com/390760/75784536-3d977f80-5d5a-11ea-94a6-49cff27f9917.png)

I'd still like to play with the animation and see if it's possible to have a less abrupt/jarring motion.

keoshi committed:
Tweak the overlay reveal/hide animation

keoshi commented:
I've pushed a small change to the overlay animation on https://github.com/Automattic/jetpack/pull/14788/commits/5b1cba98259bfebf3467c3ded26336e5f18644cb

It essentially tweaks the timing of both properties (`transform` and `opacity`) and changes the way the latter is set on the `style` param. I know this is unrelated so feel free to remove from this PR, but I thought it would be useful to see it working in this context.

keoshi commented:
One other thing I saw happening on a few of these tests is that some letters are being removed if typed very quickly. I can't say exactly when it happens, but here's a screen recording that shows the issue:

![jps-overlay-instant-type](https://user-images.githubusercontent.com/390760/75790030-7fc4bf00-5d62-11ea-9bca-6c329badfbec.gif)

Notice how the search input is `kyot` but the overlay input comes up with `ko`, which seems to happen at the very end when I was typing the last `o` indicating that `yot` was removed for some reason.

gibrown commented:
> However, I thought the overlay would only show in case search results were returned?

I think no results is still the correct result here. Like if you type in nonsense (e.g. "rsiasunoarsta"), then we should show that there are no results that match. The no results for "a" actually may be an indexing bug because it is getting stripped out because "a" is a stop word. "r" for instance is showing results.

keoshi commented:
Got it, @gibrown — thanks for clarifying!

keoshi reviewed (approved):
Other than the characters drop we talked about, this looks very good to me!
